File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    let leftWall = height[0];
    let rightWall = height[height.length - 1];
    let left = 0;
    let right = height.length - 1;
    let sumAmount = 0;

    while (left < right) {
        if (height[left] < height[right]) {
            leftWall = Math.max(height[left], leftWall)
            sumAmount += leftWall - height[left];
            left++;
        } else {
            rightWall = Math.max(height[right], rightWall)
            sumAmount += rightWall - height[right]
            right--;
        }
    }
    return sumAmount;
};

File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.go
func trap(height []int) int {
	maxWater := 0
	left := 0
	right := len(height) - 1
	leftWall := height[left]
	rightWall := height[right]

	for left < right {
		if height[left] < height[right] {
			leftWall = max(leftWall, height[left])
			maxWater += leftWall - height[left]
			left++
		} else {
			rightWall = max(rightWall, height[right])
			maxWater += rightWall - height[right]
			right--
		}
	}
	return maxWater
}